src/protocol/malicious.rs

```diff
@@ -196,7 +196,10 @@ impl<'a, F: Field> MaliciousValidator<'a, F> {
     ///
     /// ## Panics
     /// Will panic if the mutex is poisoned
-    pub async fn validate<D: DowngradeMalicious>(self, values: D) -> Result<D::Target, Error> {
+    pub async fn validate<D: DowngradeMalicious>(
+        self,
+        values: Option<D>,
+    ) -> Result<(SemiHonestContext<'a, F>, Option<D::Target>), Error> {
```

Comment:
The other option we have is for the caller of the `MaliciousValidator` to keep the original context, so that they can keep using it.  Ostensibly, we can then think of the malicious upgrade as a `.narrow()` call.  You need to validate the output to get the answer, but you are able to keep using the input context, like so:

```rust
fn do_something_complex(c: C) where C: Context {
  let v = MaliciousValidator(c.narrow(&Step::FirstMaliciousStep));
  let m_input = v.context().upgrade(input);
  let m_output = do_something_malicious(v.context(), m_input);
  let output = v.validate(m_output);

  // and do similar again, because you still own `c`...
} 
```

That saves us from passing the context through here like this - it doesn't even demand a change to the API here.

Reply:
I agree. I don't like returning the validation context... that's a rather special context.

Reply:
I don’t think it’s a problem. As with everything else - we need to ensure our code is correct :).